src/language_server_protocol.rs

```diff
@@ -793,20 +804,61 @@ impl LanguageClient {
         Ok(text.trim().into())
     }
 
+    fn edit_and_move(
+        &self,
+        (edit, cursor): (WorkspaceEdit, Option<TextDocumentPositionParams>),
+    ) -> Fallible<()> {
+        self.apply_WorkspaceEdit(&edit)?;
+
+        if let Some(cursor) = cursor {
+            self.vim()?
+                .cursor(cursor.position.line + 1, cursor.position.character + 1)?;
+        }
+
+        Ok(())
+    }
+
     fn try_handle_command_by_client(&self, cmd: &Command) -> Fallible<bool> {
+        use serde_json::{from_value, Map};
+
         if !CommandsClient.contains(&cmd.command.as_str()) {
             return Ok(false);
         }
 
-        if cmd.command == "java.apply.workspaceEdit" {
-            if let Some(ref edits) = cmd.arguments {
-                for edit in edits {
-                    let edit: WorkspaceEdit = serde_json::from_value(edit.clone())?;
-                    self.apply_WorkspaceEdit(&edit)?;
+        match cmd.command.as_str() {
+            "java.apply.workspaceEdit" => {
+                if let Some(ref edits) = cmd.arguments {
+                    for edit in edits {
+                        let edit: WorkspaceEdit = from_value(edit.clone())?;
+                        self.apply_WorkspaceEdit(&edit)?;
+                    }
                 }
             }
-        } else {
-            bail!("Not implemented: {}", cmd.command);
+            "rust-analyzer.applySourceChange" => {
+                fn json_to_workspace_edit(
+                    json: Value,
+                ) -> Fallible<(WorkspaceEdit, Option<TextDocumentPositionParams>)> {
+                    let mut map: Map<String, Value> = from_value(json)?;
+                    let ws = from_value::<WorkspaceEdit>(
+                        map.remove("workspaceEdit").unwrap_or_default(),
```

Comment:
Have you considered creating a wrap type and using serde to convert Value to concrete type?

Reply:
You mean doing the serialisation in one pass, instead of first serialising it to a `Map<String, Value>` and then separately serialising to `WorkspaceEdit` and `TextDocumentPositionParams`?

I have not. Do you maybe have some examples on how to do this? I'd gladly make the change 👍

Reply:
That's correct.

It should work similarly with other types defined in lsp-types and used in this repo. As an example, https://github.com/gluon-lang/lsp-types/blob/d78071ae83367dda76c674e392a16268fe41e843/src/lib.rs#L66-L71

The new type could be something like
```rust
struct WorkspaceEditWithCursorPosition {
    workspaceEdit: WorkspaceEdit,
    cursorPosition: TextDocumentPositionParams,
}
```

Reply:
Any updates on this?

Reply:
Sorry, not yet. It's on my list, I hope to get to it this weekend.